src/index.ts

```diff
@@ -137,54 +139,71 @@ class CruxPayPeer {
 
-    constructor(_options: ICruxPayPeerOptions) {
-        this._options = Object.assign({}, _options);
+    constructor(options: ICruxPayPeerOptions) {
+        this._options = Object.assign({}, options);
         // TODO: Need to validate options
 
-        this._getEncryptionKey = _options.getEncryptionKey;
+        this._getEncryptionKey = this._options.getEncryptionKey;
         // log.debug(`Encryption key:`, this._getEncryptionKey())
 
         // Setting up the default modules as fallbacks
         this._storage =  this._options.storage || new storage.LocalStorage();
         this._encryption = this._options.encryption || encryption.Encryption;
         this._nameService = this._options.nameService;
+        this._keyPair = this._options.keyPair;
         this.walletClientName = this._options.walletClientName;
 
         log.info(`Config mode:`, config.CONFIG_MODE);
         log.info(`CruxPayPeer Initialised`);
     }
 
     public async init() {
-        let configService;
-
+        let configService: BlockstackConfigurationService;
         if (this._hasPayIDClaimStored()) {
-            const payIDClaim = this._storage.getJSON("payIDClaim");
+            log.debug("using the stored payIDClaim");
+            const payIDClaim = (this._storage.getJSON("payIDClaim") as ICruxPayClaim);
             // log.debug(`Local payIDClaim:`, payIDClaim)
-            this._setPayIDClaim(new PayIDClaim(payIDClaim as ICruxPayClaim, { getEncryptionKey: this._getEncryptionKey }));
-            const ns: blockstackService.BlockstackService = new blockstackService.BlockstackService({domain: this.walletClientName + CRUX_DOMAIN_SUFFIX});
-            await (this._payIDClaim as PayIDClaim).decrypt();
-            await ns.restoreIdentity((this._payIDClaim as PayIDClaim).virtualAddress as string, {secrets: (this._payIDClaim as PayIDClaim).identitySecrets});
-            const status = await ns.getRegistrationStatus({secrets: (this._payIDClaim as PayIDClaim).identitySecrets});
-            await (this._payIDClaim as PayIDClaim).encrypt();
-            if (status.status === blockstackService.SubdomainRegistrationStatus.DONE) {
-                configService = new BlockstackConfigurationService(this.walletClientName, (this._payIDClaim as PayIDClaim).virtualAddress);
-            } else {
-                configService = new BlockstackConfigurationService(this.walletClientName);
+
+            // if a keyPair is provided, add a validation check on the cruxID stored
+            if (this._keyPair) {
+                const CruxIDArray = await getCruxIDByAddress(config.BLOCKSTACK.BNS_NODES[0], this.walletClientName, this._keyPair.address);
+                if (!CruxIDArray.includes(payIDClaim.virtualAddress as string)) {
+                    throw errors.ErrorHelper.getPackageError(errors.PackageErrorCode.KeyPairMismatch);
+                }
             }
+            this._setPayIDClaim(new PayIDClaim(payIDClaim, { getEncryptionKey: this._getEncryptionKey }));
```

Comment:
Why _setPayIDClaim? Did we not load it from storage only?

Reply:
this method is used to set it in memory .